hw_diag/utilities/quectel.py

```diff
@@ -0,0 +1,248 @@
+
+EG25G_DESIRED_FW = {
+    'EG25GGBR07A08M2G': 'EG25GGBR07A08M2G_30.006.30.006',
+    'EG25GGBR07A07M2G': 'EG25GGBR07A07M2G_30.003.30.003'
+}
+
+FW_STATE = '/var/data/quectel_state'
+FW_MAX_RETRIES = 5
+
+
+def sentry_capture_and_log(message):
+    capture_message(message)
+    logging.error(message)
+
+
+def find_eg25g_modem():
+    mm_proxy = ModemManager()
+    return mm_proxy.find_modem_by_properties(EG25G_UNIQUE_PROPERTIES)
+
+
+def reset_modem_and_modem_manager(modem):
+    # try if modem is still valid.
+    # a previous operation might have resulted in modem being reset/removed.
+    try:
+        modem.reset()
+    except Exception as e:
+        logging.info(f'failed to reset modem: {e}'
+                     f'modem was probably already reset/removed by previous operation')
+
+    # modem manager loses all track of the modem after reset, we need to restart it.
+    systemd_proxy = Systemd()
+    mm_unit = systemd_proxy.get_unit(DBusIds.MODEM_MANAGER_UNIT_NAME)
+    mm_restarted = mm_unit.wait_restart()
+    logging.info(f'modem manager restarted: {mm_restarted}')
+    logging.info(f'waiting {MODEM_RESET_WAIT_TIME} seconds for '
+                 f'the modem manager to pick the modem again')
+    time.sleep(MODEM_RESET_WAIT_TIME)
+
+
+def _do_upgrade(desired_fw_version):
+    systemd_proxy = Systemd()
+    mm_unit = systemd_proxy.get_unit(DBusIds.MODEM_MANAGER_UNIT_NAME)
+
+    upgrade_successful = False
+
+    stopped = mm_unit.wait_stop()
+    logging.info(f"modem manager stopped: {stopped}")
+
+    if stopped:
+        flash_cmd = ['/usr/sbin/QFirehose', '-f']
+        flash_cmd.append(f'/quectel/firmware/eg25-g/{desired_fw_version}')
+        try:
+            logging.info(f'executing {flash_cmd} to upgrade modem firmware')
+            # flash_cmd_output = subprocess.check_output(flash_cmd,
+            #                                            stderr=subprocess.STDOUT,
+            #                                            shell=True)
+            # logging.info(f'qfirehose cmd output: {flash_cmd_output}')
+            upgrade_successful = True
+        except subprocess.CalledProcessError as e:
+            logging.error(f'QFirehose failed with error: {e}')
+        except Exception as e:
+            logging.error(f'unknown error while trying to upgrade using QFirehose: {e}')
+
+    # restart the modem manager
+    mm_unit.wait_start()
+    return upgrade_successful
+
+
+def check_internet(url='https://google.com', timeout=3):
+    try:
+        urllib.request.urlopen(url, timeout=timeout)
+        return True
+    except Exception as e:
+        logging.info("failed to connect: {}".format(e))
+        return False
+
+
+def wait_check_internet(wait_time=MODEM_RESET_WAIT_TIME):
+    '''wait for wait_time seconds for modem to achieve connectivity'''
+    if check_internet():
+        return True
+
+    time.sleep(wait_time)
+    return check_internet()
+
+
+def update_setting_with_rollback(getter, setter, desired_value):
```

Comment:
The methods below here are written clearly, but their length is overwhelming. Can you break up further?

Reply:
most of the length is coming from logging and commenting. This late I am not very comfortable changing these critical functions.